libass/ass_font.c

```diff
@@ -620,3 +523,86 @@ void ass_font_clear(ASS_Font *font)
     }
     free((char *) font->desc.family.str);
 }
+
+/**
+ * \brief Convert glyph into ASS_Outline according to decoration flags
+ **/
+bool ass_get_glyph_outline(ASS_Outline *outline, int32_t *advance,
+                           FT_Face face, FT_Glyph glyph, unsigned flags)
+{
+    int32_t y_scale = face->size->metrics.y_scale;
+    int32_t adv = flags & DECO_ROTATE ? face->glyph->linearVertAdvance : glyph->advance.x;
+    *advance = adv = d16_to_d6(adv);
+
+    int n_lines = 0;
+    int32_t line_y[2][2];
+    if (adv > 0 && (flags & DECO_UNDERLINE)) {
+        TT_Postscript *ps = FT_Get_Sfnt_Table(face, ft_sfnt_post);
+        if (ps && ps->underlinePosition <= 0 && ps->underlineThickness > 0) {
+            int64_t pos  = ((int64_t) ps->underlinePosition  * y_scale + 0x8000) >> 16;
```

Comment:
FWIW this differs from `FT_MulFix`: this rounds half towards +infinity, whereas `FT_MulFix` rounds it away from zero. But we probably don’t mind.

Reply:
That's why I replaced `FT_MulFix`: its code looks like a case of heavy overengineering (they tried to combat implementation defined right shifts?), so I decided to use simple and consistent formula coupled with correct overflow handling.

Reply:
To be clear, I don’t mind this. My only lament is that `FT_MulFix(a, b)` _reads_ much nicer than `((int64_t) a * b + 0x8000) >> 16`.

---

The main point of `FT_MulFix` is outdated portability: FreeType works on platforms that don’t have native int64 (and in case the compiler emulates int64 but is bad at it, they provide faster assembler implementations).

I assume the half-away-from-zero rounding is a conscious choice; indeed, I would normally do the same.

I don’t see any difference in overflow handling though. Do you want to elaborate?

(And I’m still not a fan of assuming arithmetic right shift, but it’s a given for this PR.)

Reply:
There's a possibility of [silent truncation](https://gitlab.freedesktop.org/freetype/freetype/-/blob/master/src/base/ftcalc.c#L243) in case of `sizeof(long) == 4`.

Reply:
Ah. Right. This goes together with my other comment: I didn’t notice that you’re storing the result in `int64_t`.

Good.

<i>(Ironically, you’ll also notice that the implementation you linked does in fact assume arithmetic right shift, although the pure-32-bit-C implementation refuses to do it.)</i>